### ISSUE TITLE
Eval harness batch decoding attention mask

### ISSUE THREAD

**Issue by iankur** [pytorch/torchtune#1237]
Should we be returning proper attention mask [here](https://github.com/pytorch/torchtune/blob/f0a15c5597340efc7ad9e67b4770b6693547d143/recipes/eleuther_eval.py#L117) which will be required for batch decoding?
```python
return x, torch.ones_like(x)  # return 'mask' b/c it's expected by the harness
```

**Comment by joecummings** [comment]
Hi @iankur  - good question! We integrate with Eleuther for our evaluation, therefore we have to fit certain API contracts. Eleuther has a pretty strong integration with Hugging Face transformers and therefore most of their APIs are fit to the params that their models expect.

In this case, Hugging Face models and therefore Eleuther expect an attention mask. However, since we actually handle calling the model forward, we don't need this mask. It definitely would be best practices though :)

**Comment by iankur** [comment]
I could not find where do we mask the left padding done ([here](https://github.com/pytorch/torchtune/blob/898670f0eb58f956b5228e5a55ccac4ea0efaff8/recipes/eleuther_eval.py#L108)) in case of batch decoding. If we are masking then could you point where it is being done? If we are indeed not masking, then shouldn't we? Also, I see lm eval numbers improve by 0.5-1% on old open llm leaderboard tasks if I change batch size from 4 to 1.

**Comment by joecummings** [comment]
@iankur On a closer look at our casual generation mask, I see that it is indeed broken for bsz > 1. I will work on fixing this, but as it's not a straightforward change, I've called this out in our eleuther script and followed an issue #1250
